keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,1056 @@
+
+#### Modify VolumeGroup with existing PVCs
+
+We can add an existing PVC to the group or remove a PVC from the group without deleting it. A VOLUMEGROUP_ADD_REMOVE_EXISTING_VOLUME capability will be added to CSI Spec. Only CSI drivers supporting both VOLUMEGROUP and VOLUMEGROUP_ADD_REMOVE_EXISTING_VOLUME capabilities can support the volume group this way.
+* Create a new empty VolumeGroup.
+* Add an existing PVC to an existing VolumeGroup (VolumeGroup can be empty to start with or it can have other PVCs already) by adding VolumeGroup name to the PVC Spec.
+* VolumeGroup is modified so the existing PVC is added to the PVCList in the Status.
+  * Note: The VolumeGroup controller will be implemented to have a desired state
+    of the world and an actual state of the world. The desired state of the world
+    contains VolumeGroups with the desired PVCList while the actual state of the
+    world contains VolumeGroups with the actual PVCList. The controller will try
+    to reconcile the two by handling adding and removing multiple PVCs through a
+    single CSI RPC call each time.
+* External-provisioner will be modified so that modifying PVC by adding VolumeGroupName will trigger a ModifyVolume call (a new CSI controller RPC) to CSI driver.
+* VolumeGroup controller will be triggered to update the VolumeGroup Status.
+* Deleting a PVC from a VolumeGroup will trigger external-provisioner and the VolumeGroup controller as well.
+
+#### Create VolumeGroup from VolumeGroupSnapshot
+
+Creating a new volume group from an existing group snapshot is supported if the CSI driver supports VOLUMEGROUP capability. As a result, PVCs will be created from source snapshots and placed in a new volume group.
+
+#### Create VolumeGroup with Selector
+
+Creating a new volume group and adding existing PVCs matching the label selector to the group is supported if the CSI driver supports VOLUMEGROUP capability.
+
+CSI drivers that do not have a volume_group_id concept can use the VolumeGroup name stored in Kubernetes API server as the volume_group_id.
+
+### Create VolumeGroupSnapshot
+
+A VolumeGroupSnapshot can be created with a VolumeGroup as the source if the CSI driver supports the GROUPSNAPSHOT capability.
+* Create a VolumeGroupSnapshot with a VolumeGroup as the source.
+* This will trigger the VolumeGroupSnapshot controller to call the CreateVolumeGroupSnapshot CSI function and also create multiple VolumeSnapshot API objects with VolumeGroupSnapshot name parameter in each VolumeSnapshot Spec. This will trigger the creation of VolumeSnapshotContent API objects and a call to the CreateSnapshot CSI function.
+* When CSI driver receives CreateSnapshot request for individual snapshots with a VolumeGroupSnapshot name:
+  * Case 1: If it knows how to create a group snapshot on the storage system, it returns (nil, nil), and leave its to the CreateVolumeGroupSnapshot function to handle the snapshot creation.
```

Comment:
To find out if snapshot is part of a group, CreateSnapshot() call will need to get VolumeSnapshotGroup name. This requires change to external snapshotter to pass group name to the CreateSnapshot() call. The KEP does not mention that this change in external snapshotter is required.

Reply:
Sure.  I'll add that.  Thanks.

Reply:
Done